integration-tests/test_gcs.py

```diff
@@ -0,0 +1,120 @@
+    with smart_open.open(key, write_mode, encoding=encoding, **kwargs) as fout:
+        fout.write(content)
+    with smart_open.open(key, read_mode, encoding=encoding, **kwargs) as fin:
+        actual = fin.read()
+    return actual
+
+
+def read_length_prefixed_messages(key, read_mode, encoding=None, **kwargs):
+    with smart_open.open(key, read_mode, encoding=encoding, **kwargs) as fin:
+        actual = b''
+        length_byte = fin.read(1)
+        while len(length_byte):
+            actual += length_byte
+            msg = fin.read(ord(length_byte))
+            actual += msg
+            length_byte = fin.read(1)
+    return actual
+
+
+def test_gcs_readwrite_text(benchmark):
+    initialize_bucket()
+
+    key = _GCS_URL + '/sanity.txt'
+    text = 'с гранатою в кармане, с чекою в руке'
+    actual = benchmark(write_read, key, text, 'w', 'r', 'utf-8')
+    assert actual == text
+
+
+def test_gcs_readwrite_text_gzip(benchmark):
+    initialize_bucket()
+
+    key = _GCS_URL + '/sanity.txt.gz'
+    text = 'не чайки здесь запели на знакомом языке'
+    actual = benchmark(write_read, key, text, 'w', 'r', 'utf-8')
+    assert actual == text
+
+
+def test_gcs_readwrite_binary(benchmark):
+    initialize_bucket()
+
+    key = _GCS_URL + '/sanity.txt'
+    binary = b'this is a test'
+    actual = benchmark(write_read, key, binary, 'wb', 'rb')
+    assert actual == binary
+
+
+def test_gcs_readwrite_binary_gzip(benchmark):
+    initialize_bucket()
+
+    key = _GCS_URL + '/sanity.txt.gz'
+    binary = b'this is a test'
+    actual = benchmark(write_read, key, binary, 'wb', 'rb')
+    assert actual == binary
+
+
+def test_gcs_performance(benchmark):
+    initialize_bucket()
+
+    one_megabyte = io.BytesIO()
+    for _ in range(1024*128):
+        one_megabyte.write(b'01234567')
+    one_megabyte = one_megabyte.getvalue()
+
+    key = _GCS_URL + '/performance.txt'
+    actual = benchmark(write_read, key, one_megabyte, 'wb', 'rb')
+    assert actual == one_megabyte
+
+
+def test_gcs_performance_gz(benchmark):
+    initialize_bucket()
+
+    one_megabyte = io.BytesIO()
+    for _ in range(1024*128):
+        one_megabyte.write(b'01234567')
+    one_megabyte = one_megabyte.getvalue()
+
+    key = _GCS_URL + '/performance.txt.gz'
+    actual = benchmark(write_read, key, one_megabyte, 'wb', 'rb')
+    assert actual == one_megabyte
+
+
+def test_gcs_performance_small_reads(benchmark):
+    initialize_bucket()
+
+    ONE_MIB = 1024**2
+    one_megabyte_of_msgs = io.BytesIO()
+    msg = b'\x0f' + b'0123456789abcde'  # a length-prefixed "message"
+    for _ in range(0, ONE_MIB, len(msg)):
+        one_megabyte_of_msgs.write(msg)
+    one_megabyte_of_msgs = one_megabyte_of_msgs.getvalue()
+
+    key = _GCS_URL + '/many_reads_performance.bin'
+
+    with smart_open.open(key, 'wb') as fout:
+        fout.write(one_megabyte_of_msgs)
+
+    actual = benchmark(read_length_prefixed_messages, key, 'rb', buffering=ONE_MIB)
```

Comment:
> Shouldn't buffering be wrapped in a transport_kwargs dict? It's a GCS-only parameter, right?

Buffering is a kwarg for the built-in `open` function ([source](https://python-reference.readthedocs.io/en/latest/docs/functions/open.html)) and I thought it would be simpler for it to be named the same instead of making it specific to GCS. If you feel otherwise I can wrap it in transport_kwargs.

Reply:
> Did you try running these tests against a real GCS bucket?

Yes, I have ran both the integration and unit tests against a real bucket and they are currently passing.